src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs

```diff
@@ -2115,6 +2115,13 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
                     type = null;
                     nullability = new NullabilityInfo(CodeAnalysis.NullableAnnotation.NotAnnotated, CodeAnalysis.NullableFlowState.NotNull);
                 }
+                else if (highestBoundExpr is BoundConvertedSwitchExpression e)
+                {
+                    type = e.NaturalTypeOpt;
```

Comment:
>type = e.NaturalTypeOpt; [](start = 20, length = 24)

What is the ```boundExpr``` when we get here?

Reply:
When `highestBoundExpr is `BoundConvertedSwitchExpression` and we get here, I would expect `boundExpr` is the same bound node. I do not have confidence that the converse is true.

Reply:
> When highestBoundExpr isBoundConvertedSwitchExpressionand we get here, I would expectboundExpr` is the same bound node. I do not have confidence that the converse is true.

Then I think the code should be written in terms of ```boundExpr``` because this is what the question was asked about. I can be convinced otherwise if you can provide an example of a scenario where this wouldn't work with an explanation why that would be the case.


---
In reply to: [439585325](https://github.com/dotnet/roslyn/pull/45069#discussion_r439585325) [](ancestors = 439585325)

Reply:
Alternatively, consider adding a check that ```highestBoundExpr == boundExpr``` to the if condition above.

---
In reply to: [439590351](https://github.com/dotnet/roslyn/pull/45069#discussion_r439590351) [](ancestors = 439590351,439585325)

Reply:
I will add an assertion that the two are the same at this point (the `==` test would be redundant).

---
In reply to: [439602189](https://github.com/dotnet/roslyn/pull/45069#discussion_r439602189) [](ancestors = 439602189,439590351,439585325)

Reply:
> I will add an assertion that the two are the same at this point (the == test would be redundant).

Assert is not good enough, it doesn't affect the behavior. Please either add the condition, or provide an explanation why the code is correct when condition is false.

---
In reply to: [439610745](https://github.com/dotnet/roslyn/pull/45069#discussion_r439610745) [](ancestors = 439610745,439602189,439590351,439585325)

Reply:
The code is correct when the condition is false because the condition cannot be false.

---
In reply to: [439619189](https://github.com/dotnet/roslyn/pull/45069#discussion_r439619189) [](ancestors = 439619189,439610745,439602189,439590351,439585325)

Reply:
> because the condition cannot be false

I would prefer us to check for that rather than assume that this assumption is correct.

---
In reply to: [439622450](https://github.com/dotnet/roslyn/pull/45069#discussion_r439622450) [](ancestors = 439622450,439619189,439610745,439602189,439590351,439585325)

Reply:
If the assumption is incorrect, then adding that condition would be incorrect.

Reply:
> If the assumption is incorrect, then adding that condition would be incorrect.

Running the code would be incorrect as well, we shouldn't run it if the assumption is incorrect.

---
In reply to: [439673364](https://github.com/dotnet/roslyn/pull/45069#discussion_r439673364) [](ancestors = 439673364)

Reply:
Please provide a scenario demonstrating how it is incorrect.

---
In reply to: [439674318](https://github.com/dotnet/roslyn/pull/45069#discussion_r439674318) [](ancestors = 439674318,439673364)